### PR TITLE
Inbound batches

### LINKED ISSUE #363: Inbound: multiple batches not working
"Unable to add multiple batches of same medication when receiving inbound shipments"
the entries on the form change but then revert when editing batches on inbounds

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #363 

The issue is that when you add a batch, and this is added to the `draftInbounds` array - the update methods tied to the input columns aren't updated until they re-render. The `DataTable` is very cleverly and deliberately not re-rendering unchanged rows, to speed up the UI and make things look nice.

I wasn't able to tell the table to specifically re-render, so went with a ham-fisted approach of setting local state to indicate that a row is being added, and using that to re-render the whole table. It makes the add function a little janky - but it also makes it work.

## Tests
- [ ] Edit an inbound line. Add a batch. adjust the quantity of the original batch: expected result - the newly added batch does not disappear
- [ ] Same as above really.. but try editing the new batch and clicking `OK`: expected result - should save the values

### CODE REVIEW

File: client/packages/common/src/ui/layout/tables/DataTable.tsx [hunk 1]
```diff
@@ -19,7 +19,7 @@ import { RecordWithId } from '@common/types';
 import { useTranslation } from '@common/intl';
 import { useTableStore } from './context';
 
-export const DataTableComponent = <T extends RecordWithId>({
+const DataTableComponent = <T extends RecordWithId>({
```
Comment (mark-prins) [inline client/packages/common/src/ui/layout/tables/DataTable.tsx:22]: export isn't required. nor was this change.. 🤷

File: client/packages/common/src/ui/layout/tables/hooks/useColumns/useColumns.tsx [hunk 2]
```diff
@@ -157,15 +157,14 @@ export const createColumns = <T extends RecordWithId>(
     .addColumns(columnsToCreate)
     .build();
 
-  return columnDefinitions.map(column => {
-    return createColumnWithDefaults(column, options);
-  });
+  return columnDefinitions.map(column =>
```
Comment (mark-prins) [inline client/packages/common/src/ui/layout/tables/hooks/useColumns/useColumns.tsx:160]: just making the code a little slimmer

File: client/packages/invoices/src/InboundShipment/DetailView/modals/InboundLineEdit/InboundLineEdit.tsx [hunk 3]
```diff
@@ -19,17 +11,15 @@ import {
...
   useKeyboardHeightAdjustment,
 } from '@openmsupply-client/common';
 import { ItemRowFragment } from '@openmsupply-client/system';
-import { InboundLineEditPanel } from './InboundLineEditPanel';
-import { QuantityTable, PricingTable, LocationTable } from './TabTables';
 import { InboundLineEditForm } from './InboundLineEditForm';
 import { useInbound } from '../../../api';
 import { DraftInboundLine } from '../../../../types';
 import { CreateDraft } from '../utils';
+import { TabLayout } from './TabLayout';
```
Comment (mark-prins) [inline client/packages/invoices/src/InboundShipment/DetailView/modals/InboundLineEdit/InboundLineEdit.tsx:22]: I thought that the jsx in this file was getting too large, so lifted it out to a new component

File: client/packages/invoices/src/InboundShipment/DetailView/modals/InboundLineEdit/InboundLineEditPanel.tsx [hunk 4]
```diff
@@ -15,14 +13,11 @@ const StyledTabContainer = styled(Box)(({ theme }) => ({
...
   value: string;
-  lines: DraftInboundLine[];
-  updateDraftLine: (patch: Partial<DraftInboundLine> & { id: string }) => void;
 }
 
-export const InboundLineEditPanel: FC<PropsWithChildren<InboundLineEditPanel>> = ({
-  value,
-  children,
-}) => {
+export const InboundLineEditPanel: FC<
```
Comment (mark-prins) [inline client/packages/invoices/src/InboundShipment/DetailView/modals/InboundLineEdit/InboundLineEditPanel.tsx:18]: didn't see the need for these props, so took them out

File: client/packages/invoices/src/InboundShipment/DetailView/modals/InboundLineEdit/TabLayout.tsx [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  const [isAdding, setIsAdding] = useState(false);
+  const isMediumScreen = useIsMediumScreen();
+  const t = useTranslation('replenishment');
+
+  const onAdd = () => {
+    setIsAdding(true);
+    addDraftLine();
+  };
+
+  useEffect(() => {
```
Comment (mark-prins) [inline client/packages/invoices/src/InboundShipment/DetailView/modals/InboundLineEdit/TabLayout.tsx:49]: this is the nub of it - set the flag when adding, and have an effect set it back again when react hits that point in the render cycle

File: client/packages/invoices/src/InboundShipment/DetailView/modals/InboundLineEdit/TabLayout.tsx [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+            color="primary"
+            variant="outlined"
+            onClick={onAdd}
+            label={`${t('label.add-batch')} (+)`}
+            Icon={<PlusCircleIcon />}
+          />
+        </Box>
+      </Box>
+      {isAdding ? (
+        <Box style={{ height: 415 }} />
```
Comment (mark-prins) [inline client/packages/invoices/src/InboundShipment/DetailView/modals/InboundLineEdit/TabLayout.tsx:102]: added a fixed height to stop the modal jumping up and down in size. looks a little smoother this way

**Comment by mark-prins** [comment]
@andreievg yeah, thanks - that would do it. have run into a similar issue with URL query updates and realised that the state update issue is the problem here ( though it doesn't help with the URL query )

**Review by andreievg — APPROVED** [review]
Thanks for the changes, i had a brief look at the code refactor (not very in depth look)

**Pull request merged**